In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
import seaborn as sns

In [2]:
%matplotlib inline
pd.set_option('display.max_columns', None)

In [3]:
# import excel file November2012.xlsx
tn_2012=pd.read_excel('../data/2012_election/November2012.xlsx',
                      usecols=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE','BName1','Tally1','BName2','Tally2',
                               'BName3','Tally3','BName4','Tally4','BName5','Tally5','BName6','Tally6','BName7','Tally7',
                               'BName8','Tally8','BName9','Tally9','BName10','Tally10'])
tn_2012

COUNTY       PRECINCT                                   OFFICENAME  \
0      ANDERSON  Andersonville                      United States President   
1      ANDERSON  Andersonville                         United States Senate   
2      ANDERSON  Andersonville                         United States Senate   
3      ANDERSON  Andersonville                         United States Senate   
4      ANDERSON  Andersonville       U.S. House of Representatives District   
...         ...            ...                                          ...   
14087    WILSON    Provisional                         United States Senate   
14088    WILSON    Provisional                         United States Senate   
14089    WILSON    Provisional       U.S. House of Representatives District   
14090    WILSON    Provisional  Tennessee House of Representatives District   
14091    WILSON    Provisional  Tennessee House of Representatives District   

              ELECTDATE                       BName1  Tally1  \
0      November 6, 2012          . Mitt Romney - (R)    1110   
1      November 6, 2012           . Bob Corker - (R)    1067   
2      November 6, 2012      . Write-In - Larry Crim       0   
3      November 6, 2012  . Write-In - Mary D. Wright       0   
4      November 6, 2012    . Chuck Fleischmann - (R)     996   
...                 ...                          ...     ...   
14087  November 6, 2012      . Write-In - Larry Crim       0   
14088  November 6, 2012  . Write-In - Mary D. Wright       0   
14089  November 6, 2012          . Diane Black - (R)      36   
14090  November 6, 2012         . Mark A. Pody - (R)      16   
14091  November 6, 2012           . Susan Lynn - (R)      30   

                             BName2  Tally2                       BName3  \
0              . Barack Obama - (D)     267         . Virgil Goode - (C)   
1           . Mark E. Clayton - (D)     216         . Kermit Steck - (C)   
2      . Write-In - Gary Gene Davis       0  . Write-In - Allen Hoenicke   
3                               NaN       0                          NaN   
4          . Mary M. Headrick - (D)     282     . Matthew Deniston - (I)   
...                             ...     ...                          ...   
14087  . Write-In - Gary Gene Davis       0  . Write-In - Allen Hoenicke   
14088                           NaN       0                          NaN   
14089             . Pat Riley - (G)       9        . Scott Beasley - (I)   
14090                           NaN       0                          NaN   
14091                           NaN       0                          NaN   

       Tally3                           BName4  Tally4  \
0           3               . Jill Stein - (G)       0   
1          17          . Martin Pleasant - (G)      11   
2           0   . Write-In - De Wayne A. Jones       0   
3           0                              NaN       0   
4          48        . Write-In - Richard Hall       0   
...       ...                              ...     ...   
14087       0   . Write-In - De Wayne A. Jones       0   
14088       0                              NaN       0   
14089       9  . Write-In - Mary Amanda Keifer       0   
14090       0                              NaN       0   
14091       0                              NaN       0   

                                     BName5  Tally5  \
0          . Ross C. "Rocky" Anderson - (I)       0   
1                  . Shaun E. Crowell - (I)      18   
2                 . Write-In - Jacob Maurer       0   
3                                       NaN       0   
4      . Write-In - Lady J Jean Howard-Hill       0   
...                                     ...     ...   
14087             . Write-In - Jacob Maurer       0   
14088                                   NaN       0   
14089          . Write-In - Rachel Robinson       0   
14090                                   NaN       0   
14091                                   NaN       0   

                    

In [4]:
tn_2012.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14092 entries, 0 to 14091
Data columns (total 24 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   COUNTY      14092 non-null  object
 1   PRECINCT    14092 non-null  object
 2   OFFICENAME  14092 non-null  object
 3   ELECTDATE   14092 non-null  object
 4   BName1      14092 non-null  object
 5   Tally1      14092 non-null  int64 
 6   BName2      10660 non-null  object
 7   Tally2      14092 non-null  int64 
 8   BName3      8449 non-null   object
 9   Tally3      14092 non-null  int64 
 10  BName4      8207 non-null   object
 11  Tally4      14092 non-null  int64 
 12  BName5      7762 non-null   object
 13  Tally5      14092 non-null  int64 
 14  BName6      7059 non-null   object
 15  Tally6      14092 non-null  int64 
 16  BName7      6489 non-null   object
 17  Tally7      14092 non-null  int64 
 18  BName8      4210 non-null   object
 19  Tally8      14092 non-null  int64 
 20  BName9

In [5]:
# merge entries in all RNAME columns into new column "name"
tn_2012_namemelt=tn_2012.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['BName1','BName2','BName3','BName4','BName5','BName6','BName7','BName8','BName9','BName10'],value_name="candidate")
# drop column "variable" created by .melt
tn_2012_namemelt=tn_2012_namemelt.drop(['variable'],axis=1)                          
tn_2012_namemelt.head()

COUNTY       PRECINCT                              OFFICENAME  \
0  ANDERSON  Andersonville                 United States President   
1  ANDERSON  Andersonville                    United States Senate   
2  ANDERSON  Andersonville                    United States Senate   
3  ANDERSON  Andersonville                    United States Senate   
4  ANDERSON  Andersonville  U.S. House of Representatives District   

          ELECTDATE                    candidate  
0  November 6, 2012          . Mitt Romney - (R)  
1  November 6, 2012           . Bob Corker - (R)  
2  November 6, 2012      . Write-In - Larry Crim  
3  November 6, 2012  . Write-In - Mary D. Wright  
4  November 6, 2012    . Chuck Fleischmann - (R)

In [6]:
# merge entries in all TALLY columns into new column "vote_count"
tn_2012_pvtallymelt=tn_2012.melt(id_vars=['COUNTY','PRECINCT','OFFICENAME','ELECTDATE'],value_vars=['Tally1','Tally2','Tally3','Tally4','Tally5','Tally6','Tally7','Tally8','Tally9','Tally10'],value_name="vote_count")
# drop column "variable" created by .melt
tn_2012_pvtallymelt=tn_2012_pvtallymelt.drop(['variable'],axis=1)
tn_2012_pvtallymelt.head()

COUNTY       PRECINCT                              OFFICENAME  \
0  ANDERSON  Andersonville                 United States President   
1  ANDERSON  Andersonville                    United States Senate   
2  ANDERSON  Andersonville                    United States Senate   
3  ANDERSON  Andersonville                    United States Senate   
4  ANDERSON  Andersonville  U.S. House of Representatives District   

          ELECTDATE  vote_count  
0  November 6, 2012        1110  
1  November 6, 2012        1067  
2  November 6, 2012           0  
3  November 6, 2012           0  
4  November 6, 2012         996

In [7]:
# add "vote_count" to tn_2016_namemelt
tally=tn_2012_pvtallymelt['vote_count']
tn_2012_namemelt=tn_2012_namemelt.join(tally)
tn_2012_namemelt.head(3)

COUNTY       PRECINCT               OFFICENAME         ELECTDATE  \
0  ANDERSON  Andersonville  United States President  November 6, 2012   
1  ANDERSON  Andersonville     United States Senate  November 6, 2012   
2  ANDERSON  Andersonville     United States Senate  November 6, 2012   

                 candidate  vote_count  
0      . Mitt Romney - (R)        1110  
1       . Bob Corker - (R)        1067  
2  . Write-In - Larry Crim           0

In [8]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   COUNTY      140920 non-null  object
 1   PRECINCT    140920 non-null  object
 2   OFFICENAME  140920 non-null  object
 3   ELECTDATE   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


In [9]:
# lower case column names
tn_2012_namemelt.columns=tn_2012_namemelt.columns.str.lower()
tn_2012_namemelt.head(2)

county       precinct               officename         electdate  \
0  ANDERSON  Andersonville  United States President  November 6, 2012   
1  ANDERSON  Andersonville     United States Senate  November 6, 2012   

             candidate  vote_count  
0  . Mitt Romney - (R)        1110  
1   . Bob Corker - (R)        1067

In [10]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140920 entries, 0 to 140919
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   county      140920 non-null  object
 1   precinct    140920 non-null  object
 2   officename  140920 non-null  object
 3   electdate   140920 non-null  object
 4   candidate   75348 non-null   object
 5   vote_count  140920 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 6.5+ MB


#### 75348 non-null names  out of 140920 enties. Confirm all are 0 votes and drop rows with no candidate listed

In [11]:
# Create dataframe "blank_candidate_2012" of null valuea in "candidate" of tn_2012_namemelt
blank_candidate_2012=tn_2012_namemelt[tn_2012_namemelt['candidate'].isna()]
blank_candidate_2012

county       precinct                                   officename  \
14095   ANDERSON  Andersonville                         United States Senate   
14101   ANDERSON     Briceville                         United States Senate   
14107   ANDERSON       Bull Run                         United States Senate   
14113   ANDERSON        Clinton                         United States Senate   
14119   ANDERSON   Clinton High                         United States Senate   
...          ...            ...                                          ...   
140913    WILSON    Provisional                      United States President   
140916    WILSON    Provisional                         United States Senate   
140917    WILSON    Provisional       U.S. House of Representatives District   
140918    WILSON    Provisional  Tennessee House of Representatives District   
140919    WILSON    Provisional  Tennessee House of Representatives District   

               electdate candidate  vote_count  
14095   November 6, 2012       NaN           0  
14101   November 6, 2012       NaN           0  
14107   November 6, 2012       NaN           0  
14113   November 6, 2012       NaN           0  
14119   November 6, 2012       NaN           0  
...                  ...       ...         ...  
140913  November 6, 2012       NaN           0  
140916  November 6, 2012       NaN           0  
140917  November 6, 2012       NaN           0  
140918  November 6, 2012       NaN           0  
140919  November 6, 2012       NaN           0  

[65572 rows x 6 columns]

In [12]:
# confirm all are 0 counts
blank_candidate_2012['vote_count'].value_counts()

0    65572
Name: vote_count, dtype: int64

In [13]:
# drop all null candidates
tn_2012_namemelt=tn_2012_namemelt.dropna().reset_index(drop=True)
tn_2012_namemelt

county       precinct                              officename  \
0      ANDERSON  Andersonville                 United States President   
1      ANDERSON  Andersonville                    United States Senate   
2      ANDERSON  Andersonville                    United States Senate   
3      ANDERSON  Andersonville                    United States Senate   
4      ANDERSON  Andersonville  U.S. House of Representatives District   
...         ...            ...                                     ...   
75343    WILSON           25-2                    United States Senate   
75344    WILSON       Absentee                    United States Senate   
75345    WILSON       Absentee                    United States Senate   
75346    WILSON    Provisional                    United States Senate   
75347    WILSON    Provisional                    United States Senate   

              electdate                      candidate  vote_count  
0      November 6, 2012            . Mitt Romney - (R)        1110  
1      November 6, 2012             . Bob Corker - (R)        1067  
2      November 6, 2012        . Write-In - Larry Crim           0  
3      November 6, 2012    . Write-In - Mary D. Wright           0  
4      November 6, 2012      . Chuck Fleischmann - (R)         996  
...                 ...                            ...         ...  
75343  November 6, 2012  . Write-In - Angelia Stinnett           0  
75344  November 6, 2012  . Write-In - Fred R. Anderson           0  
75345  November 6, 2012  . Write-In - Angelia Stinnett           0  
75346  November 6, 2012  . Write-In - Fred R. Anderson           0  
75347  November 6, 2012  . Write-In - Angelia Stinnett           0  

[75348 rows x 6 columns]

In [14]:
# confirn all null candidates dropped
tn_2012_namemelt[tn_2012_namemelt['candidate'].isna()].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   county      0 non-null      object
 1   precinct    0 non-null      object
 2   officename  0 non-null      object
 3   electdate   0 non-null      object
 4   candidate   0 non-null      object
 5   vote_count  0 non-null      int64 
dtypes: int64(1), object(5)
memory usage: 0.0+ bytes


In [15]:
tn_2012_namemelt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75348 entries, 0 to 75347
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   county      75348 non-null  object
 1   precinct    75348 non-null  object
 2   officename  75348 non-null  object
 3   electdate   75348 non-null  object
 4   candidate   75348 non-null  object
 5   vote_count  75348 non-null  int64 
dtypes: int64(1), object(5)
memory usage: 3.4+ MB


### Look at cadidates with 0 votes, drop

In [16]:
tn_2012_namemelt_0votes=tn_2012_namemelt[tn_2012_namemelt['vote_count'] == 0]
tn_2012_namemelt_0votes

county       precinct            officename         electdate  \
2      ANDERSON  Andersonville  United States Senate  November 6, 2012   
3      ANDERSON  Andersonville  United States Senate  November 6, 2012   
8      ANDERSON     Briceville  United States Senate  November 6, 2012   
9      ANDERSON     Briceville  United States Senate  November 6, 2012   
14     ANDERSON       Bull Run  United States Senate  November 6, 2012   
...         ...            ...                   ...               ...   
75343    WILSON           25-2  United States Senate  November 6, 2012   
75344    WILSON       Absentee  United States Senate  November 6, 2012   
75345    WILSON       Absentee  United States Senate  November 6, 2012   
75346    WILSON    Provisional  United States Senate  November 6, 2012   
75347    WILSON    Provisional  United States Senate  November 6, 2012   

                           candidate  vote_count  
2            . Write-In - Larry Crim           0  
3        . Write-In - Mary D. Wright           0  
8            . Write-In - Larry Crim           0  
9        . Write-In - Mary D. Wright           0  
14           . Write-In - Larry Crim           0  
...                              ...         ...  
75343  . Write-In - Angelia Stinnett           0  
75344  . Write-In - Fred R. Anderson           0  
75345  . Write-In - Angelia Stinnett           0  
75346  . Write-In - Fred R. Anderson           0  
75347  . Write-In - Angelia Stinnett           0  

[31928 rows x 6 columns]

In [17]:
# drop rows with 0 in "vote_count" column
tn_2012_namemelt=tn_2012_namemelt[tn_2012_namemelt['vote_count'] != 0].reset_index(drop=True)
tn_2012_namemelt

county       precinct                                   officename  \
0        ANDERSON  Andersonville                      United States President   
1        ANDERSON  Andersonville                         United States Senate   
2        ANDERSON  Andersonville       U.S. House of Representatives District   
3        ANDERSON  Andersonville  Tennessee House of Representatives District   
4        ANDERSON     Briceville                      United States President   
...           ...            ...                                          ...   
43415  WILLIAMSON           12-1                         United States Senate   
43416  WILLIAMSON           12-2                         United States Senate   
43417  WILLIAMSON           12-3                         United States Senate   
43418      WILSON           16-1                         United States Senate   
43419      WILSON           22-1                         United States Senate   

              electdate                      candidate  vote_count  
0      November 6, 2012            . Mitt Romney - (R)        1110  
1      November 6, 2012             . Bob Corker - (R)        1067  
2      November 6, 2012      . Chuck Fleischmann - (R)         996  
3      November 6, 2012          . Dennis Powers - (R)         998  
4      November 6, 2012            . Mitt Romney - (R)         227  
...                 ...                            ...         ...  
43415  November 6, 2012  . Write-In - Angelia Stinnett           2  
43416  November 6, 2012  . Write-In - Angelia Stinnett          13  
43417  November 6, 2012  . Write-In - Angelia Stinnett           3  
43418  November 6, 2012  . Write-In - Angelia Stinnett           1  
43419  November 6, 2012  . Write-In - Angelia Stinnett           1  

[43420 rows x 6 columns]

In [18]:
tn_2012_namemelt['vote_count'].value_counts()

1       4472
2       3531
3       2819
4       2166
5       1669
        ... 
1766       1
2611       1
1670       1
1638       1
4839       1
Name: vote_count, Length: 2260, dtype: int64

In [19]:
# group county precincts into county per candidate, in descending order of votes within county
tn_2012_int=tn_2012_namemelt.groupby(['electdate','officename','county','candidate'], as_index=False)['vote_count'].sum()
tn_2012_int=tn_2012_int.sort_values(by=['county','vote_count'],ascending=False).reset_index(drop=True)
tn_2012_int.head(30)

electdate                                   officename      county  \
0   November 6, 2012                         United States Senate      WILSON   
1   November 6, 2012                      United States President      WILSON   
2   November 6, 2012       U.S. House of Representatives District      WILSON   
3   November 6, 2012  Tennessee House of Representatives District      WILSON   
4   November 6, 2012                      United States President      WILSON   
5   November 6, 2012  Tennessee House of Representatives District      WILSON   
6   November 6, 2012                         United States Senate      WILSON   
7   November 6, 2012       U.S. House of Representatives District      WILSON   
8   November 6, 2012       U.S. House of Representatives District      WILSON   
9   November 6, 2012                         United States Senate      WILSON   
10  November 6, 2012                         United States Senate      WILSON   
11  November 6, 2012                         United States Senate      WILSON   
12  November 6, 2012                      United States President      WILSON   
13  November 6, 2012                         United States Senate      WILSON   
14  November 6, 2012                         United States Senate      WILSON   
15  November 6, 2012                         United States Senate      WILSON   
16  November 6, 2012                         United States Senate      WILSON   
17  November 6, 2012                      United States President      WILSON   
18  November 6, 2012                      United States President      WILSON   
19  November 6, 2012                      United States President      WILSON   
20  November 6, 2012                      United States President      WILSON   
21  November 6, 2012       U.S. House of Representatives District      WILSON   
22  November 6, 2012                         United States Senate      WILSON   
23  November 6, 2012                         United States Senate      WILSON   
24  November 6, 2012                         United States Senate      WILSON   
25  November 6, 2012                         United States Senate      WILSON   
26  November 6, 2012                         United States Senate  WILLIAMSON   
27  November 6, 2012       U.S. House of Representatives District  WILLIAMSON   
28  November 6, 2012                      United States President  WILLIAMSON   
29  November 6, 2012                      United States President  WILLIAMSON   

                             candidate  vote_count  
0                   . Bob Corker - (R)       36993  
1                  . Mitt Romney - (R)       36109  
2                  . Diane Black - (R)       35659  
3                   . Susan Lynn - (R)       24116  
4                 . Barack Obama - (D)       14695  
5                 . Mark A. Pody - (R)       12731  
6              . Mark E. Clayton - (D)       10497  
7                . Scott Beasley - (I)        7101  
8                    . Pat Riley - (G)        4474  
9              . Martin Pleasant - (G)         749  
10                . Kermit Steck - (C)         512  
11            . Shaun E. Crowell - (I)         450  
12                . Gary Johnson - (I)         381  
13        . Troy Stephen Scoggin - (I)         176  
14                . James Higdon - (I)         136  
15          . Michel Joseph Long - (I)         131  
16              . David Gatchell - (I)         121  
17                . Virgil Goode - (C)         117  
18                  . Jill Stein - (G)         113  
19    . Ross C. "Rocky" Anderson - (I)          43  
20               . Merlin Miller - (I)          41  
21        . Write-In - Rachel Robinson          13  
22           . Write-In - Jacob Maurer           7  
23             . Write-In - Larry Crim           7  
24  . Write-In - Christopher Schappert           3  
25       . Write-In - Angelia Stinnett           2  
26                  . Bob Corker - (R)       72402  
27            .

In [20]:
# Capitalize couty names
tn_2012_int['county']=tn_2012_int['county'].str.title()
tn_2012_int.head(5)

electdate                                   officename  county  \
0  November 6, 2012                         United States Senate  Wilson   
1  November 6, 2012                      United States President  Wilson   
2  November 6, 2012       U.S. House of Representatives District  Wilson   
3  November 6, 2012  Tennessee House of Representatives District  Wilson   
4  November 6, 2012                      United States President  Wilson   

              candidate  vote_count  
0    . Bob Corker - (R)       36993  
1   . Mitt Romney - (R)       36109  
2   . Diane Black - (R)       35659  
3    . Susan Lynn - (R)       24116  
4  . Barack Obama - (D)       14695

In [21]:
tn_2012_int.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2445 entries, 0 to 2444
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   electdate   2445 non-null   object
 1   officename  2445 non-null   object
 2   county      2445 non-null   object
 3   candidate   2445 non-null   object
 4   vote_count  2445 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 95.6+ KB


In [22]:
tn_2012_int['officename'].unique()

array(['United States Senate', 'United States President',
       'U.S. House of Representatives District',
       'Tennessee House of Representatives District',
       'Tennessee Senate District'], dtype=object)

### Clean up names and greate separate party column

In [23]:
# new dataframe new_names of non-write-in candidates
new_names=tn_2012_int[tn_2012_int['candidate'].str.contains("Write")==False]
new_names

electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2435  November 6, 2012                      United States President  Anderson   
2436  November 6, 2012                         United States Senate  Anderson   
2437  November 6, 2012                         United States Senate  Anderson   
2438  November 6, 2012                      United States President  Anderson   
2439  November 6, 2012                      United States President  Anderson   

                             candidate  vote_count  
0                   . Bob Corker - (R)       36993  
1                  . Mitt Romney - (R)       36109  
2                  . Diane Black - (R)       35659  
3                   . Susan Lynn - (R)       24116  
4                 . Barack Obama - (D)       14695  
...                                ...         ...  
2435                . Jill Stein - (G)          88  
2436      . Troy Stephen Scoggin - (I)          72  
2437            . David Gatchell - (I)          60  
2438  . Ross C. "Rocky" Anderson - (I)          54  
2439             . Merlin Miller - (I)          25  

[2311 rows x 5 columns]

In [24]:
#drop period in front of names
new_names['candidate']=new_names['candidate'].str[1:]
new_names

/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2435  November 6, 2012                      United States President  Anderson   
2436  November 6, 2012                         United States Senate  Anderson   
2437  November 6, 2012                         United States Senate  Anderson   
2438  November 6, 2012                      United States President  Anderson   
2439  November 6, 2012                      United States President  Anderson   

                            candidate  vote_count  
0                    Bob Corker - (R)       36993  
1                   Mitt Romney - (R)       36109  
2                   Diane Black - (R)       35659  
3                    Susan Lynn - (R)       24116  
4                  Barack Obama - (D)       14695  
...                               ...         ...  
2435                 Jill Stein - (G)          88  
2436       Troy Stephen Scoggin - (I)          72  
2437             David Gatchell - (I)          60  
2438   Ross C. "Rocky" Anderson - (I)          54  
2439              Merlin Miller - (I)          25  

[2311 rows x 5 columns]

In [25]:
# split name fromm party desination
new_names=new_names['candidate'].str.split(pat='-',expand=True)
new_names

0     1
0                    Bob Corker    (R)
1                   Mitt Romney    (R)
2                   Diane Black    (R)
3                    Susan Lynn    (R)
4                  Barack Obama    (D)
...                          ...   ...
2435                 Jill Stein    (G)
2436       Troy Stephen Scoggin    (I)
2437             David Gatchell    (I)
2438   Ross C. "Rocky" Anderson    (I)
2439              Merlin Miller    (I)

[2311 rows x 2 columns]

In [26]:
new_names.rename(columns={0:'name',1:'party'},inplace=True)
new_names

name party
0                    Bob Corker    (R)
1                   Mitt Romney    (R)
2                   Diane Black    (R)
3                    Susan Lynn    (R)
4                  Barack Obama    (D)
...                          ...   ...
2435                 Jill Stein    (G)
2436       Troy Stephen Scoggin    (I)
2437             David Gatchell    (I)
2438   Ross C. "Rocky" Anderson    (I)
2439              Merlin Miller    (I)

[2311 rows x 2 columns]

In [27]:
# new dataframe tn_2012_int_wi of only write-in candidates
tn_2012_int_wi=tn_2012_int[tn_2012_int['candidate'].str.contains("\(")==False]
tn_2012_int_wi

electdate                              officename    county  \
21    November 6, 2012  U.S. House of Representatives District    Wilson   
22    November 6, 2012                    United States Senate    Wilson   
23    November 6, 2012                    United States Senate    Wilson   
24    November 6, 2012                    United States Senate    Wilson   
25    November 6, 2012                    United States Senate    Wilson   
...                ...                                     ...       ...   
2440  November 6, 2012  U.S. House of Representatives District  Anderson   
2441  November 6, 2012                    United States Senate  Anderson   
2442  November 6, 2012                    United States Senate  Anderson   
2443  November 6, 2012                    United States Senate  Anderson   
2444  November 6, 2012                    United States Senate  Anderson   

                               candidate  vote_count  
21          . Write-In - Rachel Robinson          13  
22             . Write-In - Jacob Maurer           7  
23               . Write-In - Larry Crim           7  
24    . Write-In - Christopher Schappert           3  
25         . Write-In - Angelia Stinnett           2  
...                                  ...         ...  
2440           . Write-In - Richard Hall           5  
2441           . Write-In - Jacob Maurer           5  
2442            . Write-In - Jim Maynard           2  
2443       . Write-In - Angelia Stinnett           1  
2444             . Write-In - Larry Crim           1  

[134 rows x 5 columns]

In [28]:
# split name fromm 'Write-in'
new_wi_name=tn_2012_int_wi['candidate'].str.split(pat='-',n=2,expand=True)
new_wi_name

0    1                       2
21    . Write  In          Rachel Robinson
22    . Write  In             Jacob Maurer
23    . Write  In               Larry Crim
24    . Write  In    Christopher Schappert
25    . Write  In         Angelia Stinnett
...       ...  ...                     ...
2440  . Write  In             Richard Hall
2441  . Write  In             Jacob Maurer
2442  . Write  In              Jim Maynard
2443  . Write  In         Angelia Stinnett
2444  . Write  In               Larry Crim

[134 rows x 3 columns]

In [29]:
# create new 'party' columnn with "Write-In"
new_wi_name['party']=new_wi_name[[0,1]].agg('-'.join,axis=1).str[2:]
new_wi_name

0    1                       2      party
21    . Write  In          Rachel Robinson  Write-In 
22    . Write  In             Jacob Maurer  Write-In 
23    . Write  In               Larry Crim  Write-In 
24    . Write  In    Christopher Schappert  Write-In 
25    . Write  In         Angelia Stinnett  Write-In 
...       ...  ...                     ...        ...
2440  . Write  In             Richard Hall  Write-In 
2441  . Write  In             Jacob Maurer  Write-In 
2442  . Write  In              Jim Maynard  Write-In 
2443  . Write  In         Angelia Stinnett  Write-In 
2444  . Write  In               Larry Crim  Write-In 

[134 rows x 4 columns]

In [30]:
#drop columns 0 and 1, rename column 2 to"candidate"
new_wi_name=new_wi_name.drop([0,1],axis=1)
new_wi_name

2      party
21           Rachel Robinson  Write-In 
22              Jacob Maurer  Write-In 
23                Larry Crim  Write-In 
24     Christopher Schappert  Write-In 
25          Angelia Stinnett  Write-In 
...                      ...        ...
2440            Richard Hall  Write-In 
2441            Jacob Maurer  Write-In 
2442             Jim Maynard  Write-In 
2443        Angelia Stinnett  Write-In 
2444              Larry Crim  Write-In 

[134 rows x 2 columns]

In [31]:
new_wi_name.rename(columns={2:'name'},inplace=True)
new_wi_name

name      party
21           Rachel Robinson  Write-In 
22              Jacob Maurer  Write-In 
23                Larry Crim  Write-In 
24     Christopher Schappert  Write-In 
25          Angelia Stinnett  Write-In 
...                      ...        ...
2440            Richard Hall  Write-In 
2441            Jacob Maurer  Write-In 
2442             Jim Maynard  Write-In 
2443        Angelia Stinnett  Write-In 
2444              Larry Crim  Write-In 

[134 rows x 2 columns]

In [32]:
# join new_names and new_wi_names
frames=[new_names,new_wi_name]
cleaned_names=pd.concat(frames)
cleaned_names

name      party
0           Bob Corker         (R)
1          Mitt Romney         (R)
2          Diane Black         (R)
3           Susan Lynn         (R)
4         Barack Obama         (D)
...                 ...        ...
2440       Richard Hall  Write-In 
2441       Jacob Maurer  Write-In 
2442        Jim Maynard  Write-In 
2443   Angelia Stinnett  Write-In 
2444         Larry Crim  Write-In 

[2445 rows x 2 columns]

### Join cleaned names and party column to tn_2012_int

In [33]:
name=cleaned_names['name']
party=cleaned_names['party']
tn_2012_final=tn_2012_int.join(name).join(party)
tn_2012_final

electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2440  November 6, 2012       U.S. House of Representatives District  Anderson   
2441  November 6, 2012                         United States Senate  Anderson   
2442  November 6, 2012                         United States Senate  Anderson   
2443  November 6, 2012                         United States Senate  Anderson   
2444  November 6, 2012                         United States Senate  Anderson   

                          candidate  vote_count               name      party  
0                . Bob Corker - (R)       36993        Bob Corker         (R)  
1               . Mitt Romney - (R)       36109       Mitt Romney         (R)  
2               . Diane Black - (R)       35659       Diane Black         (R)  
3                . Susan Lynn - (R)       24116        Susan Lynn         (R)  
4              . Barack Obama - (D)       14695      Barack Obama         (D)  
...                             ...         ...                ...        ...  
2440      . Write-In - Richard Hall           5       Richard Hall  Write-In   
2441      . Write-In - Jacob Maurer           5       Jacob Maurer  Write-In   
2442       . Write-In - Jim Maynard           2        Jim Maynard  Write-In   
2443  . Write-In - Angelia Stinnett           1   Angelia Stinnett  Write-In   
2444        . Write-In - Larry Crim           1         Larry Crim  Write-In   

[2445 rows x 7 columns]

In [34]:
tn_2012_final['party'].value_counts()

 (I)         940
 (R)         504
 (D)         421
 (G)         256
 (C)         190
Write-In     134
Name: party, dtype: int64

In [35]:
# change "(R)" to "Republican" and "(D)"to "Democratic" in "party" column
tn_2012_final["party"]=tn_2012_final['party'].str.replace('\(R\)',"Republican").str.replace('\(D\)',"Democratic")
tn_2012_final

electdate                                   officename    county  \
0     November 6, 2012                         United States Senate    Wilson   
1     November 6, 2012                      United States President    Wilson   
2     November 6, 2012       U.S. House of Representatives District    Wilson   
3     November 6, 2012  Tennessee House of Representatives District    Wilson   
4     November 6, 2012                      United States President    Wilson   
...                ...                                          ...       ...   
2440  November 6, 2012       U.S. House of Representatives District  Anderson   
2441  November 6, 2012                         United States Senate  Anderson   
2442  November 6, 2012                         United States Senate  Anderson   
2443  November 6, 2012                         United States Senate  Anderson   
2444  November 6, 2012                         United States Senate  Anderson   

                          candidate  vote_count               name  \
0                . Bob Corker - (R)       36993        Bob Corker    
1               . Mitt Romney - (R)       36109       Mitt Romney    
2               . Diane Black - (R)       35659       Diane Black    
3                . Susan Lynn - (R)       24116        Susan Lynn    
4              . Barack Obama - (D)       14695      Barack Obama    
...                             ...         ...                ...   
2440      . Write-In - Richard Hall           5       Richard Hall   
2441      . Write-In - Jacob Maurer           5       Jacob Maurer   
2442       . Write-In - Jim Maynard           2        Jim Maynard   
2443  . Write-In - Angelia Stinnett           1   Angelia Stinnett   
2444        . Write-In - Larry Crim           1         Larry Crim   

            party  
0      Republican  
1      Republican  
2      Republican  
3      Republican  
4      Democratic  
...           ...  
2440    Write-In   
2441    Write-In   
2442    Write-In   
2443    Write-In   
2444    Write-In   

[2445 rows x 7 columns]

In [36]:
# create data frame tn_2012_uss for senatorial election
tn_2012_uss=tn_2012_final[tn_2012_final['officename'].str.match('United States Senate')]
tn_2012_uss=tn_2012_uss.reset_index(drop=True)
tn_2012_uss

electdate            officename    county  \
0    November 6, 2012  United States Senate    Wilson   
1    November 6, 2012  United States Senate    Wilson   
2    November 6, 2012  United States Senate    Wilson   
3    November 6, 2012  United States Senate    Wilson   
4    November 6, 2012  United States Senate    Wilson   
..                ...                   ...       ...   
955  November 6, 2012  United States Senate  Anderson   
956  November 6, 2012  United States Senate  Anderson   
957  November 6, 2012  United States Senate  Anderson   
958  November 6, 2012  United States Senate  Anderson   
959  November 6, 2012  United States Senate  Anderson   

                         candidate  vote_count                name  \
0               . Bob Corker - (R)       36993         Bob Corker    
1          . Mark E. Clayton - (D)       10497    Mark E. Clayton    
2          . Martin Pleasant - (G)         749    Martin Pleasant    
3             . Kermit Steck - (C)         512       Kermit Steck    
4         . Shaun E. Crowell - (I)         450   Shaun E. Crowell    
..                             ...         ...                 ...   
955         . David Gatchell - (I)          60     David Gatchell    
956      . Write-In - Jacob Maurer           5        Jacob Maurer   
957       . Write-In - Jim Maynard           2         Jim Maynard   
958  . Write-In - Angelia Stinnett           1    Angelia Stinnett   
959        . Write-In - Larry Crim           1          Larry Crim   

           party  
0     Republican  
1     Democratic  
2            (G)  
3            (C)  
4            (I)  
..           ...  
955          (I)  
956    Write-In   
957    Write-In   
958    Write-In   
959    Write-In   

[960 rows x 7 columns]

In [37]:
tn_2012_uss.head(20)

electdate            officename      county  \
0   November 6, 2012  United States Senate      Wilson   
1   November 6, 2012  United States Senate      Wilson   
2   November 6, 2012  United States Senate      Wilson   
3   November 6, 2012  United States Senate      Wilson   
4   November 6, 2012  United States Senate      Wilson   
5   November 6, 2012  United States Senate      Wilson   
6   November 6, 2012  United States Senate      Wilson   
7   November 6, 2012  United States Senate      Wilson   
8   November 6, 2012  United States Senate      Wilson   
9   November 6, 2012  United States Senate      Wilson   
10  November 6, 2012  United States Senate      Wilson   
11  November 6, 2012  United States Senate      Wilson   
12  November 6, 2012  United States Senate      Wilson   
13  November 6, 2012  United States Senate  Williamson   
14  November 6, 2012  United States Senate  Williamson   
15  November 6, 2012  United States Senate  Williamson   
16  November 6, 2012  United States Senate  Williamson   
17  November 6, 2012  United States Senate  Williamson   
18  November 6, 2012  United States Senate  Williamson   
19  November 6, 2012  United States Senate  Williamson   

                             candidate  vote_count                    name  \
0                   . Bob Corker - (R)       36993             Bob Corker    
1              . Mark E. Clayton - (D)       10497        Mark E. Clayton    
2              . Martin Pleasant - (G)         749        Martin Pleasant    
3                 . Kermit Steck - (C)         512           Kermit Steck    
4             . Shaun E. Crowell - (I)         450       Shaun E. Crowell    
5         . Troy Stephen Scoggin - (I)         176   Troy Stephen Scoggin    
6                 . James Higdon - (I)         136           James Higdon    
7           . Michel Joseph Long - (I)         131     Michel Joseph Long    
8               . David Gatchell - (I)         121         David Gatchell    
9            . Write-In - Jacob Maurer           7            Jacob Maurer   
10             . Write-In - Larry Crim           7              Larry Crim   
11  . Write-In - Christopher Schappert           3   Christopher Schappert   
12       . Write-In - Angelia Stinnett           2        Angelia Stinnett   
13                  . Bob Corker - (R)       72402             Bob Corker    
14             . Mark E. Clayton - (D)       16789        Mark E. Clayton    
15             . Martin Pleasant - (G)        1632        Martin Pleasant    
16            . Shaun E. Crowell - (I)        1209       Shaun E. Crowell    
17                . Kermit Steck - (C)         801           Kermit Steck    
18        . Troy Stephen Scoggin - (I)         235   Troy Stephen Scoggin    
19                . James Higdon - (I)         203           James Higdon    

          party  
0    Republican  
1    Democratic  
2           (G)  
3           (C)  
4           (I)  
5           (I)  
6           (I)  
7           (I)  
8           (I)  
9     Write-In   
10    Write-In   
11    Write-In   
12    Write-In   
13   Republican  
14   Democratic  
15          (G)  
16          (I)  
17          (C)  
18          (I)  
19          (I)

### Add column that has total vote count by count for use in calculating county vote percentages

In [38]:
# create new dataframe "county_vote" to use to find sum of votes per county
county_vote=tn_2012_uss.loc[0:,['county','vote_count']]
county_vote

county  vote_count
0      Wilson       36993
1      Wilson       10497
2      Wilson         749
3      Wilson         512
4      Wilson         450
..        ...         ...
955  Anderson          60
956  Anderson           5
957  Anderson           2
958  Anderson           1
959  Anderson           1

[960 rows x 2 columns]

In [39]:
#sum county votes
county_vote_sum=county_vote.groupby(['county'])['vote_count'].sum()
county_vote_sum

county
Anderson      27962
Bedford       13207
Benton         5977
Bledsoe        4170
Blount        46677
              ...  
Wayne          4934
Weakley       11851
White          8622
Williamson    93761
Wilson        49784
Name: vote_count, Length: 95, dtype: int64

In [40]:
# use "county_vote_sum" to add column "sum_county_vote" to tn_2016_pres correlating with value in each row
tn_2012_uss['sum_county_vote']=tn_2012_uss['county'].map(county_vote_sum)
tn_2012_uss.head(20)

electdate            officename      county  \
0   November 6, 2012  United States Senate      Wilson   
1   November 6, 2012  United States Senate      Wilson   
2   November 6, 2012  United States Senate      Wilson   
3   November 6, 2012  United States Senate      Wilson   
4   November 6, 2012  United States Senate      Wilson   
5   November 6, 2012  United States Senate      Wilson   
6   November 6, 2012  United States Senate      Wilson   
7   November 6, 2012  United States Senate      Wilson   
8   November 6, 2012  United States Senate      Wilson   
9   November 6, 2012  United States Senate      Wilson   
10  November 6, 2012  United States Senate      Wilson   
11  November 6, 2012  United States Senate      Wilson   
12  November 6, 2012  United States Senate      Wilson   
13  November 6, 2012  United States Senate  Williamson   
14  November 6, 2012  United States Senate  Williamson   
15  November 6, 2012  United States Senate  Williamson   
16  November 6, 2012  United States Senate  Williamson   
17  November 6, 2012  United States Senate  Williamson   
18  November 6, 2012  United States Senate  Williamson   
19  November 6, 2012  United States Senate  Williamson   

                             candidate  vote_count                    name  \
0                   . Bob Corker - (R)       36993             Bob Corker    
1              . Mark E. Clayton - (D)       10497        Mark E. Clayton    
2              . Martin Pleasant - (G)         749        Martin Pleasant    
3                 . Kermit Steck - (C)         512           Kermit Steck    
4             . Shaun E. Crowell - (I)         450       Shaun E. Crowell    
5         . Troy Stephen Scoggin - (I)         176   Troy Stephen Scoggin    
6                 . James Higdon - (I)         136           James Higdon    
7           . Michel Joseph Long - (I)         131     Michel Joseph Long    
8               . David Gatchell - (I)         121         David Gatchell    
9            . Write-In - Jacob Maurer           7            Jacob Maurer   
10             . Write-In - Larry Crim           7              Larry Crim   
11  . Write-In - Christopher Schappert           3   Christopher Schappert   
12       . Write-In - Angelia Stinnett           2        Angelia Stinnett   
13                  . Bob Corker - (R)       72402             Bob Corker    
14             . Mark E. Clayton - (D)       16789        Mark E. Clayton    
15             . Martin Pleasant - (G)        1632        Martin Pleasant    
16            . Shaun E. Crowell - (I)        1209       Shaun E. Crowell    
17                . Kermit Steck - (C)         801           Kermit Steck    
18        . Troy Stephen Scoggin - (I)         235   Troy Stephen Scoggin    
19                . James Higdon - (I)         203           James Higdon    

          party  sum_county_vote  
0    Republican            49784  
1    Democratic            49784  
2           (G)            49784  
3           (C)            49784  
4           (I)            49784  
5           (I)            49784  
6           (I)            49784  
7           (I)            49784  
8           (I)            49784  
9     Write-In             49784  
10    Write-In             49784  
11    Write-In             49784  
12    Write-In             49784  
13   Republican            93761  
14   Democratic            93761  
15          (G)            93761  
16          (I)            93761  
17          (C)            93761  
18          (I)            93761  
19          (I)            93761

In [41]:
# create "pct_county_vote" column for tn_2016_pres
tn_2012_uss['pct_county_vote']=round(tn_2012_uss['vote_count']/tn_2012_uss['sum_county_vote']*100,2)
tn_2012_uss.head(20)

electdate            officename      county  \
0   November 6, 2012  United States Senate      Wilson   
1   November 6, 2012  United States Senate      Wilson   
2   November 6, 2012  United States Senate      Wilson   
3   November 6, 2012  United States Senate      Wilson   
4   November 6, 2012  United States Senate      Wilson   
5   November 6, 2012  United States Senate      Wilson   
6   November 6, 2012  United States Senate      Wilson   
7   November 6, 2012  United States Senate      Wilson   
8   November 6, 2012  United States Senate      Wilson   
9   November 6, 2012  United States Senate      Wilson   
10  November 6, 2012  United States Senate      Wilson   
11  November 6, 2012  United States Senate      Wilson   
12  November 6, 2012  United States Senate      Wilson   
13  November 6, 2012  United States Senate  Williamson   
14  November 6, 2012  United States Senate  Williamson   
15  November 6, 2012  United States Senate  Williamson   
16  November 6, 2012  United States Senate  Williamson   
17  November 6, 2012  United States Senate  Williamson   
18  November 6, 2012  United States Senate  Williamson   
19  November 6, 2012  United States Senate  Williamson   

                             candidate  vote_count                    name  \
0                   . Bob Corker - (R)       36993             Bob Corker    
1              . Mark E. Clayton - (D)       10497        Mark E. Clayton    
2              . Martin Pleasant - (G)         749        Martin Pleasant    
3                 . Kermit Steck - (C)         512           Kermit Steck    
4             . Shaun E. Crowell - (I)         450       Shaun E. Crowell    
5         . Troy Stephen Scoggin - (I)         176   Troy Stephen Scoggin    
6                 . James Higdon - (I)         136           James Higdon    
7           . Michel Joseph Long - (I)         131     Michel Joseph Long    
8               . David Gatchell - (I)         121         David Gatchell    
9            . Write-In - Jacob Maurer           7            Jacob Maurer   
10             . Write-In - Larry Crim           7              Larry Crim   
11  . Write-In - Christopher Schappert           3   Christopher Schappert   
12       . Write-In - Angelia Stinnett           2        Angelia Stinnett   
13                  . Bob Corker - (R)       72402             Bob Corker    
14             . Mark E. Clayton - (D)       16789        Mark E. Clayton    
15             . Martin Pleasant - (G)        1632        Martin Pleasant    
16            . Shaun E. Crowell - (I)        1209       Shaun E. Crowell    
17                . Kermit Steck - (C)         801           Kermit Steck    
18        . Troy Stephen Scoggin - (I)         235   Troy Stephen Scoggin    
19                . James Higdon - (I)         203           James Higdon    

          party  sum_county_vote  pct_county_vote  
0    Republican            49784            74.31  
1    Democratic            49784            21.09  
2           (G)            49784             1.50  
3           (C)            49784             1.03  
4           (I)            49784             0.90  
5           (I)            49784             0.35  
6           (I)            49784             0.27  
7           (I)            49784             0.26  
8           (I)            49784             0.24  
9     Write-In             49784             0.01  
10    Write-In             49784             0.01  
11    Write-In             49784             0.01  
12    Write-In             49784             0.00  
13   Republican            93761            77.22  
14   Democratic            93761            17.91  
15          (G)            93761             1.74  
16          (I)            93761             1.29  
17          (C)            93761             0.85  
18          (I)            93761             0.25  
19          (I)            93761             0.22

In [42]:
# how many parties?
tn_2012_uss['party'].value_counts()

 (I)           475
Write-In       105
 (G)            95
 Republican     95
 (C)            95
 Democratic     95
Name: party, dtype: int64

## Create dataframe "tn_2012_county_pres_pct"  with percent values per party per county

In [43]:
# create new dataframe with reduced columns values only
tn_2012_pct_county=tn_2012_uss.loc[0:,['electdate','county','party','pct_county_vote']]
tn_2012_pct_county

electdate    county        party  pct_county_vote
0    November 6, 2012    Wilson   Republican            74.31
1    November 6, 2012    Wilson   Democratic            21.09
2    November 6, 2012    Wilson          (G)             1.50
3    November 6, 2012    Wilson          (C)             1.03
4    November 6, 2012    Wilson          (I)             0.90
..                ...       ...          ...              ...
955  November 6, 2012  Anderson          (I)             0.21
956  November 6, 2012  Anderson    Write-In              0.02
957  November 6, 2012  Anderson    Write-In              0.01
958  November 6, 2012  Anderson    Write-In              0.00
959  November 6, 2012  Anderson    Write-In              0.00

[960 rows x 4 columns]

In [44]:
# create new dataframe with Republican values only
tn_2012_rep_pct=tn_2012_pct_county[tn_2012_pct_county['party']== " Republican"].reset_index(drop=True)
tn_2012_rep_pct

electdate      county        party  pct_county_vote
0   November 6, 2012      Wilson   Republican            74.31
1   November 6, 2012  Williamson   Republican            77.22
2   November 6, 2012       White   Republican            70.04
3   November 6, 2012     Weakley   Republican            67.60
4   November 6, 2012       Wayne   Republican            78.86
..               ...         ...          ...              ...
90  November 6, 2012      Blount   Republican            76.80
91  November 6, 2012     Bledsoe   Republican            73.60
92  November 6, 2012      Benton   Republican            62.67
93  November 6, 2012     Bedford   Republican            73.23
94  November 6, 2012    Anderson   Republican            70.11

[95 rows x 4 columns]

In [45]:
# rename 'pct_county_vote' to 'rep_pct' and drop "party" column to prepare for merge 
tn_2012_rep_pct=tn_2012_rep_pct.rename(columns={'pct_county_vote':'Republican'},inplace=False).drop(['party'], axis=1)
tn_2012_rep_pct

electdate      county  Republican
0   November 6, 2012      Wilson       74.31
1   November 6, 2012  Williamson       77.22
2   November 6, 2012       White       70.04
3   November 6, 2012     Weakley       67.60
4   November 6, 2012       Wayne       78.86
..               ...         ...         ...
90  November 6, 2012      Blount       76.80
91  November 6, 2012     Bledsoe       73.60
92  November 6, 2012      Benton       62.67
93  November 6, 2012     Bedford       73.23
94  November 6, 2012    Anderson       70.11

[95 rows x 3 columns]

In [46]:
# create new dataframe with Democratic values only
tn_2012_dem_pct=tn_2012_pct_county[tn_2012_pct_county['party']== " Democratic"].reset_index(drop=True)
tn_2012_dem_pct

electdate      county        party  pct_county_vote
0   November 6, 2012      Wilson   Democratic            21.09
1   November 6, 2012  Williamson   Democratic            17.91
2   November 6, 2012       White   Democratic            24.51
3   November 6, 2012     Weakley   Democratic            27.64
4   November 6, 2012       Wayne   Democratic            17.67
..               ...         ...          ...              ...
90  November 6, 2012      Blount   Democratic            17.77
91  November 6, 2012     Bledsoe   Democratic            23.05
92  November 6, 2012      Benton   Democratic            32.52
93  November 6, 2012     Bedford   Democratic            23.09
94  November 6, 2012    Anderson   Democratic            23.95

[95 rows x 4 columns]

In [47]:
# rename 'pct_county_vote' to 'dem_pct' and drop "party" to prepare for merge 
tn_2012_dem_pct=tn_2012_dem_pct.rename(columns={'pct_county_vote':'Democratic'},inplace=False).drop(['party'], axis=1)
tn_2012_dem_pct

electdate      county  Democratic
0   November 6, 2012      Wilson       21.09
1   November 6, 2012  Williamson       17.91
2   November 6, 2012       White       24.51
3   November 6, 2012     Weakley       27.64
4   November 6, 2012       Wayne       17.67
..               ...         ...         ...
90  November 6, 2012      Blount       17.77
91  November 6, 2012     Bledsoe       23.05
92  November 6, 2012      Benton       32.52
93  November 6, 2012     Bedford       23.09
94  November 6, 2012    Anderson       23.95

[95 rows x 3 columns]

In [48]:
# create new dataframe with non Republican or Democratic values only
tn_2012_other_pct=tn_2012_pct_county[(tn_2012_pct_county['party']!=" Republican")&(tn_2012_pct_county['party']!=" Democratic")].reset_index(drop=True)
tn_2012_other_pct

electdate    county      party  pct_county_vote
0    November 6, 2012    Wilson        (G)             1.50
1    November 6, 2012    Wilson        (C)             1.03
2    November 6, 2012    Wilson        (I)             0.90
3    November 6, 2012    Wilson        (I)             0.35
4    November 6, 2012    Wilson        (I)             0.27
..                ...       ...        ...              ...
765  November 6, 2012  Anderson        (I)             0.21
766  November 6, 2012  Anderson  Write-In              0.02
767  November 6, 2012  Anderson  Write-In              0.01
768  November 6, 2012  Anderson  Write-In              0.00
769  November 6, 2012  Anderson  Write-In              0.00

[770 rows x 4 columns]

In [49]:
# sum by county
tn_2012_other_pct=tn_2012_other_pct.groupby(['county'],as_index=False)['pct_county_vote'].sum()
tn_2012_other_pct

county  pct_county_vote
0     Anderson             5.94
1      Bedford             3.69
2       Benton             4.81
3      Bledsoe             3.35
4       Blount             5.43
..         ...              ...
90       Wayne             3.46
91     Weakley             4.77
92       White             5.45
93  Williamson             4.87
94      Wilson             4.58

[95 rows x 2 columns]

In [50]:
# rename 'pct_county_vote' to 'other_pct' and drop "party" to prepare for merge 
tn_2012_other_pct=tn_2012_other_pct.rename(columns={'pct_county_vote':'Other'})
tn_2012_other_pct

county  Other
0     Anderson   5.94
1      Bedford   3.69
2       Benton   4.81
3      Bledsoe   3.35
4       Blount   5.43
..         ...    ...
90       Wayne   3.46
91     Weakley   4.77
92       White   5.45
93  Williamson   4.87
94      Wilson   4.58

[95 rows x 2 columns]

In [52]:
# join tn_2012_rep, tn_2012_dem_pct and tn_2012_other_pct to form tn_2012_county_uss_pct
dem=tn_2012_dem_pct['Democratic']
other=tn_2012_other_pct['Other']
tn_2012_county_uss_pct=tn_2012_rep_pct.join(dem).join(other)
tn_2012_county_uss_pct

electdate      county  Republican  Democratic  Other
0   November 6, 2012      Wilson       74.31       21.09   5.94
1   November 6, 2012  Williamson       77.22       17.91   3.69
2   November 6, 2012       White       70.04       24.51   4.81
3   November 6, 2012     Weakley       67.60       27.64   3.35
4   November 6, 2012       Wayne       78.86       17.67   5.43
..               ...         ...         ...         ...    ...
90  November 6, 2012      Blount       76.80       17.77   3.46
91  November 6, 2012     Bledsoe       73.60       23.05   4.77
92  November 6, 2012      Benton       62.67       32.52   5.45
93  November 6, 2012     Bedford       73.23       23.09   4.87
94  November 6, 2012    Anderson       70.11       23.95   4.58

[95 rows x 5 columns]

### Add column for Republican margin over Democrat for heat mapping in Tableau

In [53]:
tn_2012_county_uss_pct['Rep_Margin']= tn_2012_county_uss_pct['Republican']-tn_2012_county_uss_pct['Democratic']
tn_2012_county_uss_pct.head(50)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   November 6, 2012      Wilson       74.31       21.09   5.94       53.22
1   November 6, 2012  Williamson       77.22       17.91   3.69       59.31
2   November 6, 2012       White       70.04       24.51   4.81       45.53
3   November 6, 2012     Weakley       67.60       27.64   3.35       39.96
4   November 6, 2012       Wayne       78.86       17.67   5.43       61.19
5   November 6, 2012  Washington       72.98       22.07   3.11       50.91
6   November 6, 2012      Warren       67.62       27.46   4.52       40.16
7   November 6, 2012   Van Buren       65.53       30.95   5.77       34.58
8   November 6, 2012       Union       77.30       18.68   4.20       58.62
9   November 6, 2012      Unicoi       75.91       19.17   4.28       56.74
10  November 6, 2012   Trousdale       63.78       31.52   5.56       32.26
11  November 6, 2012      Tipton       73.15       23.36   3.68       49.79
12  November 6, 2012      Sumner       75.22       20.49   4.39       54.73
13  November 6, 2012    Sullivan       75.69       20.48   4.12       55.21
14  November 6, 2012     Stewart       62.72       32.41   3.85       30.31
15  November 6, 2012       Smith       68.90       26.87   4.40       42.03
16  November 6, 2012      Shelby       43.90       52.14   2.63       -8.24
17  November 6, 2012      Sevier       79.80       16.09   5.04       63.71
18  November 6, 2012  Sequatchie       76.19       20.09   6.80       56.10
19  November 6, 2012       Scott       80.24       16.39   3.96       63.85
20  November 6, 2012  Rutherford       65.50       29.31   5.12       36.19
21  November 6, 2012   Robertson       72.58       22.87   5.60       49.71
22  November 6, 2012       Roane       74.66       20.89   2.88       53.77
23  November 6, 2012        Rhea       79.78       17.03   3.35       62.75
24  November 6, 2012      Putnam       71.12       23.02   3.23       48.10
25  November 6, 2012        Polk       70.64       25.80   3.73       44.84
26  November 6, 2012     Pickett       75.39       21.32   3.31       54.07
27  November 6, 2012       Perry       66.51       28.49   4.30       38.02
28  November 6, 2012     Overton       67.78       27.86   4.13       39.92
29  November 6, 2012       Obion       66.96       26.92   4.61       40.04
30  November 6, 2012      Morgan       75.35       20.41   3.60       54.94
31  November 6, 2012       Moore       76.06       20.32   3.35       55.74
32  November 6, 2012  Montgomery       60.12       34.79   3.26       25.33
33  November 6, 2012      Monroe       74.59       20.06   4.09       54.53
34  November 6, 2012       Meigs       75.55       21.57   4.99       53.98
35  November 6, 2012     Mcnairy       71.68       25.35   3.17       46.33
36  November 6, 2012      Mcminn       78.09       18.26   3.57       59.83
37  November 6, 2012       Maury       62.69       32.12   2.51       30.57
38  November 6, 2012    Marshall       66.67       28.44   3.41       38.23
39  November 6, 2012      Marion       66.59       30.18   4.17       36.41
40  November 6, 2012     Madison       59.12       37.72   6.46       21.40
41  November 6, 2012       Macon       78.66       17.02   4.89       61.64
42  November 6, 2012      Loudon       80.08       15.41   6.26       64.67
43  November 6, 2012     Lincoln       73.30       22.55   5.92       50.75
44  November 6, 2012       Lewis       68.41       25.33   4.43       43.08
45  November 6, 2012    Lawrence       73.45       23.23   4.68       50.22
46  November 6, 2012  Lauderdale       59.44       36.51   6.72       22.93
47  November 6, 2012        Lake       55.54       39.85   4.60       15.69
48  November 6, 2012        Knox       69.89       23.38   4.07       46.51
49  November 6, 2012     Johnson       77.73       17.60   3.33       60.13

In [54]:
# correct county names: Dekalb to DeKalb, Mcminn to McMinn, Mcnairy to McNairy 
tn_2012_county_uss_pct['county']=tn_2012_county_uss_pct['county'].str.replace('Dekalb','DeKalb').str.replace('Mcminn','McMinn').str.replace('Mcnairy','McNairy')
tn_2012_county_uss_pct.head(50)

electdate      county  Republican  Democratic  Other  Rep_Margin
0   November 6, 2012      Wilson       74.31       21.09   5.94       53.22
1   November 6, 2012  Williamson       77.22       17.91   3.69       59.31
2   November 6, 2012       White       70.04       24.51   4.81       45.53
3   November 6, 2012     Weakley       67.60       27.64   3.35       39.96
4   November 6, 2012       Wayne       78.86       17.67   5.43       61.19
5   November 6, 2012  Washington       72.98       22.07   3.11       50.91
6   November 6, 2012      Warren       67.62       27.46   4.52       40.16
7   November 6, 2012   Van Buren       65.53       30.95   5.77       34.58
8   November 6, 2012       Union       77.30       18.68   4.20       58.62
9   November 6, 2012      Unicoi       75.91       19.17   4.28       56.74
10  November 6, 2012   Trousdale       63.78       31.52   5.56       32.26
11  November 6, 2012      Tipton       73.15       23.36   3.68       49.79
12  November 6, 2012      Sumner       75.22       20.49   4.39       54.73
13  November 6, 2012    Sullivan       75.69       20.48   4.12       55.21
14  November 6, 2012     Stewart       62.72       32.41   3.85       30.31
15  November 6, 2012       Smith       68.90       26.87   4.40       42.03
16  November 6, 2012      Shelby       43.90       52.14   2.63       -8.24
17  November 6, 2012      Sevier       79.80       16.09   5.04       63.71
18  November 6, 2012  Sequatchie       76.19       20.09   6.80       56.10
19  November 6, 2012       Scott       80.24       16.39   3.96       63.85
20  November 6, 2012  Rutherford       65.50       29.31   5.12       36.19
21  November 6, 2012   Robertson       72.58       22.87   5.60       49.71
22  November 6, 2012       Roane       74.66       20.89   2.88       53.77
23  November 6, 2012        Rhea       79.78       17.03   3.35       62.75
24  November 6, 2012      Putnam       71.12       23.02   3.23       48.10
25  November 6, 2012        Polk       70.64       25.80   3.73       44.84
26  November 6, 2012     Pickett       75.39       21.32   3.31       54.07
27  November 6, 2012       Perry       66.51       28.49   4.30       38.02
28  November 6, 2012     Overton       67.78       27.86   4.13       39.92
29  November 6, 2012       Obion       66.96       26.92   4.61       40.04
30  November 6, 2012      Morgan       75.35       20.41   3.60       54.94
31  November 6, 2012       Moore       76.06       20.32   3.35       55.74
32  November 6, 2012  Montgomery       60.12       34.79   3.26       25.33
33  November 6, 2012      Monroe       74.59       20.06   4.09       54.53
34  November 6, 2012       Meigs       75.55       21.57   4.99       53.98
35  November 6, 2012     McNairy       71.68       25.35   3.17       46.33
36  November 6, 2012      McMinn       78.09       18.26   3.57       59.83
37  November 6, 2012       Maury       62.69       32.12   2.51       30.57
38  November 6, 2012    Marshall       66.67       28.44   3.41       38.23
39  November 6, 2012      Marion       66.59       30.18   4.17       36.41
40  November 6, 2012     Madison       59.12       37.72   6.46       21.40
41  November 6, 2012       Macon       78.66       17.02   4.89       61.64
42  November 6, 2012      Loudon       80.08       15.41   6.26       64.67
43  November 6, 2012     Lincoln       73.30       22.55   5.92       50.75
44  November 6, 2012       Lewis       68.41       25.33   4.43       43.08
45  November 6, 2012    Lawrence       73.45       23.23   4.68       50.22
46  November 6, 2012  Lauderdale       59.44       36.51   6.72       22.93
47  November 6, 2012        Lake       55.54       39.85   4.60       15.69
48  November 6, 2012        Knox       69.89       23.38   4.07       46.51
49  November 6, 2012     Johnson       77.73       17.60   3.33       60.13

In [55]:
#read TN_county_index_shape.csv to add index from TN Counties shape file for joining in Tableau
county_index=pd.read_csv('../data/TN_county_index_shape.csv')
county_index

county  INDEX_NO
0     Anderson        79
1      Bedford        26
2       Benton        41
3      Bledsoe        31
4       Blount        52
..         ...       ...
90       Wayne         6
91     Weakley        56
92       White        67
93  Williamson        44
94      Wilson        65

[95 rows x 2 columns]

In [56]:
# join with tn_2012_county_pres_pct
tn_2012_county_uss_pct=tn_2012_county_uss_pct.merge(county_index,on="county")
tn_2012_county_uss_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  \
0   November 6, 2012      Wilson       74.31       21.09   5.94       53.22   
1   November 6, 2012  Williamson       77.22       17.91   3.69       59.31   
2   November 6, 2012       White       70.04       24.51   4.81       45.53   
3   November 6, 2012     Weakley       67.60       27.64   3.35       39.96   
4   November 6, 2012       Wayne       78.86       17.67   5.43       61.19   
..               ...         ...         ...         ...    ...         ...   
90  November 6, 2012      Blount       76.80       17.77   3.46       59.03   
91  November 6, 2012     Bledsoe       73.60       23.05   4.77       50.55   
92  November 6, 2012      Benton       62.67       32.52   5.45       30.15   
93  November 6, 2012     Bedford       73.23       23.09   4.87       50.14   
94  November 6, 2012    Anderson       70.11       23.95   4.58       46.16   

    INDEX_NO  
0         65  
1         44  
2         67  
3         56  
4          6  
..       ...  
90        52  
91        31  
92        41  
93        26  
94        79  

[95 rows x 7 columns]

In [57]:
tn_2012_county_uss_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   electdate   95 non-null     object 
 1   county      95 non-null     object 
 2   Republican  95 non-null     float64
 3   Democratic  95 non-null     float64
 4   Other       95 non-null     float64
 5   Rep_Margin  95 non-null     float64
 6   INDEX_NO    95 non-null     int64  
dtypes: float64(4), int64(1), object(2)
memory usage: 5.9+ KB


In [58]:
# covert "electdate" to datetime from object
tn_2012_county_uss_pct['electdate']=pd.to_datetime(tn_2012_county_uss_pct['electdate'],format='%B %d, %Y')
tn_2012_county_uss_pct

electdate      county  Republican  Democratic  Other  Rep_Margin  INDEX_NO
0  2012-11-06      Wilson       74.31       21.09   5.94       53.22        65
1  2012-11-06  Williamson       77.22       17.91   3.69       59.31        44
2  2012-11-06       White       70.04       24.51   4.81       45.53        67
3  2012-11-06     Weakley       67.60       27.64   3.35       39.96        56
4  2012-11-06       Wayne       78.86       17.67   5.43       61.19         6
..        ...         ...         ...         ...    ...         ...       ...
90 2012-11-06      Blount       76.80       17.77   3.46       59.03        52
91 2012-11-06     Bledsoe       73.60       23.05   4.77       50.55        31
92 2012-11-06      Benton       62.67       32.52   5.45       30.15        41
93 2012-11-06     Bedford       73.23       23.09   4.87       50.14        26
94 2012-11-06    Anderson       70.11       23.95   4.58       46.16        79

[95 rows x 7 columns]

In [59]:
tn_2012_county_uss_pct.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 95 entries, 0 to 94
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   electdate   95 non-null     datetime64[ns]
 1   county      95 non-null     object        
 2   Republican  95 non-null     float64       
 3   Democratic  95 non-null     float64       
 4   Other       95 non-null     float64       
 5   Rep_Margin  95 non-null     float64       
 6   INDEX_NO    95 non-null     int64         
dtypes: datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 5.9+ KB


In [60]:
tn_2012_county_uss_pct.to_excel(r'../data/2012_election/tn_2012_county_uss_pct.xlsx',index=False)